zebra-state/src/service/check/anchors.rs

```diff
@@ -44,24 +48,39 @@ pub(crate) fn anchors_refer_to_earlier_treestates(
         //
         // The FIRST JOINSPLIT in a transaction MUST refer to the output treestate
         // of a previous block.
+        if transaction.has_sprout_joinsplit_data() {
+            // > The anchor of each JoinSplit description in a transaction MUST refer to
+            // > either some earlier block’s final Sprout treestate, or to the interstitial
+            // > output treestate of any prior JoinSplit description in the same transaction.
+            //
+            // https://zips.z.cash/protocol/protocol.pdf#joinsplit
+            let mut interstitial_roots: HashSet<sprout::tree::Root> = HashSet::new();
+
+            let mut interstitial_note_commitment_tree = parent_chain.sprout_note_commitment_tree();
+
+            for (n, joinsplit) in transaction.sprout_groth16_joinsplits().enumerate() {
+                for cm in joinsplit.commitments {
+                    interstitial_note_commitment_tree
+                        .append(cm)
+                        .expect("note commitment should be appendable to the tree");
+                }
+
+                interstitial_roots.insert(interstitial_note_commitment_tree.root());
```

Comment:
I think this code must be moved after the check below, because

> The anchor of each JoinSplit description in a transaction MUST refer to [...] to the interstitial output treestate of any **prior** JoinSplit description in the same transaction."

So a transaction can't point to its own state, but that's currently allowed in the code.

With that change, the `n== 0` check won't be needed since `interstitial_roots` will be empty in the first iteration anyway.

(As a optimization, you can avoid computing and updating the root in the last iteration)

Reply:
Fixed. Funny (or rather not funny) that I had thought of this in the other branch, but not here.

Reply:
Just throwing the other implementation out there: https://github.com/ZcashFoundation/zebra/blob/e0a054242ab09f0e4c4944f17b89128fd3bfdc41/zebra-state/src/service/check/anchors.rs#L48